File: 2683-neighboring-bitwise-xor/2683-neighboring-bitwise-xor (sum derived).cpp
class Solution {
public:
    bool doesValidArrayExist(vector<int>& derived) {
        int n = derived.size();
        bool test0 = 0, test1 = 1, last0 = 0, last1 = 1;    // if original[0] = 0 and if original[0] = 1.
        for (int i=0; i<n-1; i++){
            last0 ^= derived[i];    // calculate original[n-1] if original[0] = 0.
            last1 ^= derived[i];    // calculate original[n-1] if original[0] = 1.
        }
        if (test0 ^ last0 == derived[n-1] || test1 ^ last1 == derived[n-1]) // if (original[0] ^ original[n-1] == derived[n-1]).
            return true;
        else
            return false;
    }
};

// Reference from LeetCode - 2683. Neighboring Bitwise XOR　解題心得, https://home.gamer.com.tw/artwork.php?sn=5970680.

File: 2683-neighboring-bitwise-xor/2683-neighboring-bitwise-xor (sum derived).py
class Solution(object):
    def doesValidArrayExist(self, derived):
        """
        :type derived: List[int]
        :rtype: bool
        """
        n = len(derived)
        sum = 0
        for i in range(n):
            sum += derived[i]
        return sum%2==0

# Reference from conner1231230's solution.

'''

If i = n - 1, then derived[i] = original[i] ⊕ original[0].
Otherwise, derived[i] = original[i] ⊕ original[i + 1].

derived[0] ⊕ derived[1] ⊕ ⋯ ⊕ derived[n−1] =
original[0] ⊕ original[1] ⊕ original[1] ⊕ original[2] ⊕ ⋯ ⊕ original[n−1] ⊕ original[0] = 
original[0] ⊕ original[0] ⊕ original[1] ⊕ original[1] ⊕ ⋯ ⊕ original[n−1] ⊕ original[n−1].

because a ⊕ a = 0,
so original[0] ⊕ original[0] ⊕ original[1] ⊕ original[1] ⊕ ⋯ ⊕ original[n−1] ⊕ original[n−1] = 0,
so derived[0] ⊕ derived[1] ⊕ ⋯ ⊕ derived[n−1] = 0,
so if 1 in derived is even, XOR of derived is 0, return true; else if 1 in derived is odd, XOR of derived is 1, return false.

'''